yacl/crypto/primitives/psu/krtw19_psu.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 zhangwfjh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/psu/krtw19_psu.h"
+
+#include <algorithm>
+#include <array>
+#include <iterator>
+#include <random>
+#include <unordered_set>
+#include <utility>
+
+#include "yacl/utils/serialize.h"
+
+namespace yacl::crypto {
+
+namespace {
+
+// reference: https://eprint.iacr.org/2019/1234.pdf (Figure 2)
+constexpr float ZETA{0.06f};
+constexpr size_t BIN_SIZE{64ul};  // m+1
+constexpr uint128_t BOT{};
+constexpr size_t NUM_BINS_PER_BATCH{16ul};
+constexpr size_t BATCH_SIZE{NUM_BINS_PER_BATCH * BIN_SIZE};
+
+constexpr size_t NUM_BASE_OT{128ul};
+constexpr size_t NUM_INKP_OT{512ul};
+
+static std::random_device rd;
+static std::mt19937 gen(rd());
+
+struct U128Hasher {
+  size_t operator()(const uint128_t& x) const {
+    auto hash = Blake3_128({&x, sizeof x});
+    size_t ret;
+    std::memcpy(&ret, &hash, 1);
```

Comment:
P0: `std::memcpy` will copy just 1 byte, but the bit width of `size_t` is not less than 16 (2 bytes). (since C99)
see: https://en.cppreference.com/w/c/types/size_t